### PR TITLE
Artifact Registry fixes

### PR DESCRIPTION
**Pull request by vladsaveliev**

### CODE REVIEW

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
This was definitely an oversight, thanks for pointing it out! I think there is a slightly deeper problem here though. Here are my thoughts. Note these are organized more in terms of my thought process rather than final recommendations. Those are at the end.

1.  `publicly_available_images` is probably a more appropriate name than `public_gcr_images` if the docker prefix is variable.
2. The `publish-public-images` Make target should take the `DOCKER_PREFIX` instead and use that instead of gcr.io + PROJECT in `publish_public_images.sh`.
3. The semantics of `public_gcr_images` is slightly different now because we've added the domain. Whether this is more or less correct is up for debate, more on that later.
4. We add the registry to `image_ref` for `HAIL_GENETICS_IMAGES` that we also keep in `gcr.io`, I assume to avoid dockerhub rate limits and maybe faster to pull from within google? This feels confusing to me, and I don't know why we have both `HAIL_GENETICS_IMAGES` and `PUBLIC_GCR_IMAGES`. It feels like there are three cases:
    1. A user gives an image with a registry/domain in which case we respect that and do nothing
    2. A user gives a hail-sponsored image without a registry, in which case we make the decision as to which registry to use
    3. A user gives an unknown image without a registry, in which case we treat it as a dockerhub image.

For an example, let's take `DOCKER_PREFIX=gcr.io/hail-vdc`.
If a user submits a job using `hailgenetics/python-dill`, what do we do? We see that `image_ref.name() in HAIL_GENETICS_IMAGES`, so the current implementation adds the domain `gcr.io` and `image_ref.path` becomes `hail-vdc/gcr.io/hailgenetics/python-dill`. Clearly that is incorrect. The proposed change would make the path `hail-vdc/hailgenetics/python-dill`. This does in fact exist! However, this doesn't feel like the intended path based on this part of `publish-public-images.sh`, which suggests it should be `hail-vdc/python-dill`:

```bash
docker=hailgenetics/$name:$hail_pip_version
google=gcr.io/$project/$name:$hail_pip_version
latest=gcr.io/$project/$name:latest
```

I will let @danking chime in on the right path forward and what semantics we want. Here's what I think we want, if we want to select registries for images that don't specify one:

1. `publish-public-images.sh` should publish to `${DOCKER_PREFIX}/hailgenetics/$name:$hail_pip_version`
2. This way we can say `HAIL_PUBLIC_IMAGES = HAIL_GENETICS_IMAGES + ['python-dill', 'query', ...]`. Then we have
```python
        if image_ref.name() in HAIL_PUBLIC_IMAGES:
            domain, path_prefix = DOCKER_PREFIX.split('/', maxsplit=1)
            image_ref.domain = domain
            image_ref.path = path_prefix + image_ref.path
            image_ref.pull_on_users_behalf = True
```
and we remove the line `is_public_gcr_image = self.image_ref.name() in PUBLIC_GCR_IMAGES` in favor of `image_ref. pull_on_users_behalf` deciding whether to use the batch worker token to pull the image.

Adding to the confusion is the tension between referring to image names as domain + path vs docker_prefix + colloquial name, since where the GCP project ends up is different between the two options. 

Of course a simple alternative is to demand image references are explicit about their registry, and follow the normal Docker convention where unspecified domains draw from dockerhub, but we've come this far haven't we 🙂 

Thoughts and alternative welcome. I hope I have overcomplicated this and there is a simpler explanation.

**Comment by danking** [comment]
> 1. publish-public-images.sh should publish to ${DOCKER_PREFIX}/hailgenetics/$name:$hail_pip_version

Yeah. My bad on that, I broke this by changing the other code to insert `hailgenetics` without changing `publish-public-images.sh`.

> 2. ...

This seems great to me.

---

I don't want to users to write `gcr.io/hail-vdc/hailgenetics/python-dill` because I think of GCR as an implementation detail of Hail Batch. Instead, I want them to believe they're pulling from `docker.io/hailgenetics/python-dill`, but with a cache in between docker hub and us.

File: build.yaml [hunk 1]
```diff
@@ -2435,8 +2435,8 @@ steps:
       export HAIL_DEFAULT_NAMESPACE={{ default_ns.name }}
       export HAIL_NETCAT_UBUNTU_IMAGE={{ netcat_ubuntu_image.image }}
       export HAIL_HAIL_BASE_IMAGE={{ hail_base_image.image }}
-      export DOCKER_PREFIX="{{ global.docker_prefix }}"
-      export DOCKER_ROOT_IMAGE="{{ global.docker_root_image }}"
+      export HAIL_DOCKER_ROOT_IMAGE="{{ global.docker_root_image }}"
+      export HAIL_DOCKER_PREFIX="{{ global.docker_prefix }}"
```
[the remark this answers is not in the document]
Comment (vladsaveliev) [reply build.yaml:2439]: Do you suggest renaming `HAIL_DOCKER_ROOT_IMAGE` and `HAIL_DOCKER_PREFIX` to `DOCKER_ROOT_IMAGE` and `DOCKER_PREFIX` everywhere?

**Comment by vladsaveliev** [comment]
> For an example, let's take DOCKER_PREFIX=gcr.io/hail-vdc.

> If a user submits a job using `hailgenetics/python-dill`, what do we do? We see that `image_ref.name()` in `HAIL_GENETICS_IMAGES`, so the current implementation adds the domain gcr.io and image_ref.path becomes `hail-vdc/gcr.io/hailgenetics/python-dill`. Clearly that is incorrect. The proposed change would make the path `hail-vdc/hailgenetics/python-dill`. This does in fact exist! However, this doesn't feel like the intended path based on this part of `publish-public-images.sh`, which suggests it should be `hail-vdc/python-dill`:

Hmm, the proposed implementation actually prints 'hail-vdc/python-dill' for `image_ref.path`:

```
image_ref = parse_docker_image_reference('hailgenetics/python-dill')
# image_ref.path == 'python-dill'

if image_ref.name() in HAIL_GENETICS_IMAGES:
    image_ref.domain = DOCKER_PREFIX.split('/', maxsplit=1)[0]  # 'gcr.io'
    image_ref.path = '/'.join(DOCKER_PREFIX.split('/')[1:] + [image_ref.path])  

# image_ref.path == 'hail-vdc/python-dill'
```

Though it should probably be `hail-vdc/hailgenetics/python-dill`? (I also just updated `publish-public-images.sh` to push to `$docker_prefix/hailgenetics/$name`).

File: ci/deployment.yaml [hunk 2]
```diff
@@ -65,6 +65,10 @@ spec:
                  key: docker_root_image
            - name: HAIL_GCP_ZONE
              value: "{{ global.zone }}"
+           - name: HAIL_DOCKER_PREFIX
+             value: "{{ global.docker_prefix }}"
+           - name: HAIL_DOCKER_ROOT_IMAGE
+             value: "{{ global.docker_root_image }}"
```
[the remark this answers is not in the document]
Comment (vladsaveliev) [reply ci/deployment.yaml]: Ah, my bad. Removed the duplicated ones.

**Comment by daniel-goldstein** [comment]
Thanks for pointing that out. The spec is ambiguous, though in this case I think we can hack around it if we just set 
```python
image_ref.path = '/'.join(DOCKER_PREFIX.split('/')[1:] + [image_ref.name()])
```
**before** setting the domain, and maybe leave a comment explaining the rewrite.

File: build.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (daniel-goldstein) [reply build.yaml:2439]: I'd likely prefer it the other way around, like you've done, but that would require changing a lot more than `build.yaml`. Which is why I would suggest leaving it alone for now (unless there's something broken currently with this naming that I'm missing?)

[the remark this answers is not in the document]
Comment (vladsaveliev) [reply build.yaml]: Gotcha! Not sure why I put the `HAIL_` prefixes into `build.yaml` in the first place. Reverted now.

**Comment by vladsaveliev** [comment]
Turns out it becomes less hacky if we add into the regex the requirement for the domain chunk to have a `.` in it. This way, for e.g. `hailgenetics/python-dill`, it doesn't try to interpret `hailgenetics` as a `domain`. But instead sets the whole bit as a `path`. Which we then basically attach to the docker prefix.

**Comment by daniel-goldstein** [comment]
>  the requirement for the domain chunk to have a . in it

I think this is a bit problematic. We could reasonably run a docker registry whose `/etc/hosts` entry on the worker is `registry`, and it would be difficult to debug if we cannot rely on docker image specifications to align to the standard. `ParsedDockerImageReference.name` is essentially just the provided docker image reference with the tag and digest stripped, so I don't think my proposed solution is much of an issue.

**Comment by vladsaveliev** [comment]
Gotcha, makes sense about `/etc/hosts`. Reverted the `.` requirement and added something similar to your proposed solution.

Though I see that the most critical bug was fixed in a separate PR: https://github.com/hail-is/hail/pull/10410
So feel free to just close this PR if needed :)

**Comment by daniel-goldstein** [comment]
Thanks @vladsaveliev, I'll authorize this so we can run it through CI and try to get it through.

**Review by daniel-goldstein — APPROVED** [review]

**Pull request merged**